src/vm/threads.cpp

```diff
@@ -6551,6 +6559,106 @@ Thread *ThreadStore::GetThreadList(Thread *cursor)
     return GetAllThreadList(cursor, (Thread::TS_Unstarted | Thread::TS_Dead), 0);
 }
 
+const LONG ThreadStore::s_DeadThreadCountThresholdForGCTrigger = 25;
+const DWORD ThreadStore::s_DeadThreadGCTriggerPeriodMilliseconds = 1000 * 60 * 10;
+
+//
+// Dead thread count for GC trigger
+// - Interlocked operations are used, as a background GC thread may also be involved in reading/writing the value
+// - The thread store lock could theoreticaly be used, but it would add new GC suspension points and it is unclear that this is
+//   ok at these points
+//
+
+LONG ThreadStore::GetDeadThreadCountForGCTrigger()
+{
+    // Usage of this value should not assume a nonnegative value, see Increment and Decrement
+    return FastInterlockCompareExchange(&m_DeadThreadCountForGCTrigger, -1, -1);
+}
+
+void ThreadStore::IncrementDeadThreadCountForGCTrigger()
+{
+    // Overflow can be ignored, as the Decrement function below will take care of it
+    FastInterlockIncrement(&m_DeadThreadCountForGCTrigger);
+}
+
+void ThreadStore::DecrementDeadThreadCountForGCTrigger()
+{
+    // Decrement to a minimum of zero, or reset the count to zero if it happened to overflow in an Increment operation
+    LONG currentCount = m_DeadThreadCountForGCTrigger;
+    while (true)
+    {
+        LONG nextCount = currentCount > 0 ? currentCount - 1 : 0;
+        LONG countBeforeUpdate =
+            FastInterlockCompareExchange(&m_DeadThreadCountForGCTrigger, nextCount, currentCount);
+        if (countBeforeUpdate == currentCount || countBeforeUpdate == 0)
+        {
+            break;
+        }
+        currentCount = countBeforeUpdate;
+    }
+}
+
+void ThreadStore::ResetDeadThreadCountForGCTrigger()
+{
+    FastInterlockExchange(&m_DeadThreadCountForGCTrigger, 0);
+}
+
+//
+// Previous max-generation GC time in milliseconds
+// - For the same reasons as for the dead thread count for GC trigger, interlocked operations are used. See above for more info.
+//
+
+DWORD ThreadStore::GetPreviousMaxGenerationGCTimeMilliseconds()
+{
+    static_assert_no_msg(sizeof(LONG) == sizeof(DWORD));
+    return static_cast<DWORD>(FastInterlockCompareExchange(&m_PreviousMaxGenerationGCTimeMilliseconds, 0, 0));
+}
+
+void ThreadStore::SetPreviousMaxGenerationGCTimeMilliseconds()
+{
```

Comment:
This function has been removed now, but the reasoning is the same as for the write to count